### PR TITLE
Closes #78 Added ADR for data ingest and k-v store structure

### LINKED ISSUE #78: Design API Contract for Data Ingest
ADR to define a generalised ingest API

- REST definition with noun/verb of how we ingest the data
- consider PartitionKey and attributes

~Dependent on the LEI schema investigation~

### PR DESCRIPTION
**Pull request by mkcomer**
Added ADR for data ingest

### CODE REVIEW

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+### Endpoint
+
+```
+POST  /csv
+// Endpoint ingests data into data reconciliation app
+	• URL Params
+        None
+	• Headers
+        Content-Type: application/json
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:27]: I think this is a file upload operation and `Content-Type` will be  `multipart/form-data`.

Also, for the security of this endpoint, are we going to request a token or an API key/secret in the headers?
Maybe an API key is simpler at this stage?

```
x-api-key: {UUID}
```

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	• Request is ccfapp.Request object
+		○ Reuest Body = CSV text
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:29]: Is it the file upload or posting CSV file content as string in the body?

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	• Status: 200
+    • Status Text: "Data has been successfully recorded"
+```
+
+### Service Layer - CSV Service
+
+```
+// injects member data into CCF K-V store
+SubmitData(userId, csv body))
+	• Parse spreadsheet
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:40]: Parse `csv`?

Also, what will be the format of the CSV? Will it include headers too?

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+POST  /csv
+// Endpoint ingests data into data reconciliation app
+	• URL Params
+        None
+	• Headers
+        Content-Type: application/json
+	• Request is ccfapp.Request object
+		○ Reuest Body = CSV text
+		○ Request UserID =
+	• Status: 200
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:31]: Do we need to design the errors in the contract too?

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    • Status Text: "Data has been successfully recorded"
+```
+
+### Service Layer - CSV Service
+
+```
+// injects member data into CCF K-V store
+SubmitData(userId, csv body))
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:39]: Everything is data :) 

Given the service is called `csvService`,  this method could be `parse(userId, body): ServiceResult<value>` and the returning data can be sent to `Repository` to save into `K-V` store.

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	• Parse spreadsheet
+		○ Any validation error
+        ○ Grab headers of spreadsheet
+		○ Check if key exists
+            - if exists, update vote
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:44]: Why updating vote or saving to `K-V` store are both the concerns of this service? Could these concerns distributed to singular places?

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            - else, save to K-V Store
+```
+
+### K-V Store data structure
+
+- Key: Unique Id (string)
+- Value: Attribute object
+  - id: attribute header (string)
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:52]: This `id` is confusing when we have `unique id` in the Key.
Is this actually `companyName`?

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  - type: string or number (string for our MVP)
+  - votes: Map<UserID (string), Attribute Value (string)>
+
+### Sample Data
+
+From Member 1:
+
+| lei   | company_name |
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:60]: Are we going to follow `snake_case`? Alternatively, `kebab-case` for the CSV headers  can be more readable?

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+- Members will upload data via csv
+- Members will upload multiple records (batch) or single records via csv
+- Data will have one attribute, and we will only work with string attributes for now.
+
+## Decision
+
+### Endpoint
+
+```
+POST  /csv
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:22]: Is this endpoint actually posting the data or voting? If that is voting, could this be called `POST /votes`?

File: data-reconciliation-app/docs/adr/01-data-ingest.md (hunk 8 above)
Comment (ross-p-smith) [reply data-reconciliation-app/docs/adr/01-data-ingest.md:60]: Personally, I have a preference on _ as mouse clicks always work better ¯\_(ツ)_/¯

Comment (ross-p-smith) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:60]: Should we remove reference to lei and just call it ID?

File: data-reconciliation-app/docs/adr/01-data-ingest.md (hunk 1 above)
Comment (Aymalla) [reply data-reconciliation-app/docs/adr/01-data-ingest.md:27]: I think we need to consider one the available authenication options provided by CCF https://microsoft.github.io/CCF/main/build_apps/auth/index.html

Comment (mkcomer) [reply data-reconciliation-app/docs/adr/01-data-ingest.md]: added to latest commit

File: data-reconciliation-app/docs/adr/01-data-ingest.md (hunk 9 above)
Comment (mkcomer) [reply data-reconciliation-app/docs/adr/01-data-ingest.md]: added to latest commit

File: data-reconciliation-app/docs/adr/01-data-ingest.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkcomer) [reply data-reconciliation-app/docs/adr/01-data-ingest.md]: added to latest commit

File: data-reconciliation-app/docs/adr/01-data-ingest.md (hunk 8 above)
Comment (mkcomer) [reply data-reconciliation-app/docs/adr/01-data-ingest.md]: added to latest commit

File: data-reconciliation-app/docs/adr/01-data-ingest.md (hunk 2 above)
Comment (mkcomer) [reply data-reconciliation-app/docs/adr/01-data-ingest.md]: will be json - added to latest commit

File: data-reconciliation-app/docs/adr/01-data-ingest.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    - Status: 401
+    - Status Text: "Unauthorized"
+  - Validation Error
+    - Status: 400
+    - Status Text: "Incorrect format"
+
+### Service
+
+```
+submitVotes(userId: string, csv: string)
```
Comment (Aymalla) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:108]: submitVotes(userId: string, data: [])

Comment (Aymalla) [inline data-reconciliation-app/docs/adr/01-data-ingest.md:108]: submitVotes(userId: string, data: [])

**Review by Aymalla — APPROVED** [review]
LGTM, just nit comment

**Review by ross-p-smith — APPROVED** [review]

**Pull request merged**